### PR TITLE
Limit number of open sessions

### LINKED ISSUE #153: Max Builder Sessions
# Rationale

https://github.com/flashbots/suave-geth/pull/115#discussion_r1431376986

### PR DESCRIPTION
**Pull request by lthibault**
## 📝 Summary

Use a semaphore channel to limit the number of open sessions. This is a hack because we're relying on session timeout behavior. Recommend ref-counting API that releases refs when block-height changes in near-term future.

Resolves #153 

## 📚 References

---

* [x] I have seen and agree to CONTRIBUTING.md

### CODE REVIEW

File: suave/builder/api/api_server.go [hunk 1]
```diff
@@ -6,25 +6,25 @@ import (
 	"github.com/ethereum/go-ethereum/core/types"
 )
 
-// sessionManager is the backend that manages the session state of the builder API.
-type sessionManager interface {
-	NewSession() (string, error)
+// SessionManager is the backend that manages the session state of the builder API.
```
Comment (lthibault) [inline suave/builder/api/api_server.go:9]: I noticed that this was an unexported type being returned from the public function `NewServer`, which is kinda awkward.  Exporting `SessionManager` is a drive-by improvement.

**Comment by lthibault** [comment]
Test passes locally ... re-running 🧐 
I think this is ready for review in the meantime, though.

**Comment by lthibault** [comment]
I'm noticing that failing test is [potentially non-deterministic](https://github.com/flashbots/suave-geth/blob/main/suave/cstore/redis_backends_test.go#L58-L63).  Did I break this, or is something flaky?

Test passes on my laptop, so I can't `git bisect` to see if I introduced this 🤔

**Comment by metachris** [comment]
nice code and comments. is there a good way to write a test ensuring it works as expected?

**Comment by lthibault** [comment]
Added unit tests.  @metachris I think it's a good idea to hook in a envvar flag here, but I'd like to do it in a separate PR because it'll affect a different set of modules.  Merging now and following up with that change.
